packages/ckeditor5-engine/src/conversion/mapper.ts

```diff
@@ -675,6 +775,456 @@ export default class Mapper extends /* #__PURE__ */ EmitterMixin() {
+	 *
+	 * Starting state:
+	 *
+	 * ```
+	 * <p>This is <strong>some <em>heavily <u>formatted</u>^ piece of</em></strong> text.</p>
+	 * ```
+	 *
+	 * Then we remove " piece of " and invalidate some cache:
+	 *
+	 * ```
+	 * <p>This is <strong>some <em>heavily <u>formatted</u>^</em></strong> text.</p>
+	 * ```
+	 *
+	 * Now, if we ask for model offset after letter "d" in "formatted", we should get a position in " text", but we will get in `<em>`.
+	 * For this scenario, we need to hoist the position.
+	 *
+	 * ```
+	 * <p>This is <strong>some <em>heavily <u>formatted</u></em></strong>^ text.</p>
+	 * ```
+	 */
+	private _hoistViewPosition( viewPosition: ViewPosition ): ViewPosition {
+		while ( viewPosition.parent.parent && !this._cachedMapping.has( viewPosition.parent as any ) && viewPosition.isAtEnd ) {
+			const parent = viewPosition.parent.parent;
+			const offset = parent.getChildIndex( viewPosition.parent ) + 1;
+
+			viewPosition = new ViewPosition( parent, offset );
+		}
+
+		return viewPosition;
+	}
+
+	/**
+	 * Starts tracking given `viewContainer`, which must be mapped to a model element or model document fragment.
+	 *
+	 * Note, that this method is automatically called by {@link ~MapperCache#get `MapperCache#getClosest()`} and there is no need to call it
+	 * manually.
+	 *
+	 * This method initializes the cache for `viewContainer` and adds callbacks for
+	 * {@link module:engine/view/node~ViewNodeChangeEvent `change` event} fired by `viewContainer`. `MapperCache` listens to `change` event
+	 * on the tracked elements to invalidate the stored cache.
+	 */
+	public startTracking( viewContainer: ViewElement | ViewDocumentFragment ): CacheItem {
+		const viewPosition = new ViewPosition( viewContainer, 0 );
+		const initialCacheItem = { viewPosition, modelOffset: 0 };
+		const initialCache: MappingCache = {
+			maxModelOffset: 0,
+			cacheList: [ initialCacheItem ],
+			cacheMap: new Map( [ [ 0, initialCacheItem ] ] )
+		};
+
+		this._cachedMapping.set( viewContainer, initialCache );
+
+		// Listen to changes in tracked view containers in order to invalidate the cache.
+		//
+		// Possible performance improvement. This event bubbles, so if there are multiple tracked (mapped) elements that are ancestors
+		// then this will be unnecessarily fired for each ancestor. This could be rewritten to listen only to roots and document fragments.
+		viewContainer.on<ViewNodeChangeEvent>( 'change:children', this._invalidateOnChildrenChangeCallback );
+		viewContainer.on<ViewNodeChangeEvent>( 'change:text', this._invalidateOnTextChangeCallback );
+
+		return initialCacheItem;
+	}
+
+	/**
+	 * Stops tracking given `viewContainer`.
+	 *
+	 * It removes the cached data and stops listening to {@link module:engine/view/node~ViewNodeChangeEvent `change` event} on the
+	 * `viewContainer`.
+	 */
+	public stopTracking( viewContainer: ViewElement | ViewDocumentFragment ): void {
+		this.stopListening( viewContainer );
```

Comment:
This probably won't do anything since `this` knows nothing about those listeners. Shouldn't it be using those callback properties?

Reply:
Of course it does nothing :facepalm:. After rewriting some of the code I forgot to use `.off` here :facepalm:.